test/deform/linearized_deform_test.py

```diff
@@ -0,0 +1,302 @@
+    return retval
+
+
+def hat_function_grad0(x, **kwargs):
+    """First component of gradient of hat function with support in square
+    with side 2r."""
+    r = kwargs.pop('r', 0.5)
+    tmp = np.where(abs(x[0]) >= 1e-6, -x[0] / abs(x[0]) * (r - abs(x[1])), 0)
+    return tmp * (abs(x[0]) < r) * (abs(x[1]) < r)
+
+
+def hat_function_grad1(x, **kwargs):
+    """2nd component of gradient of hat function with support in square
+    with side 2r."""
+    r = kwargs.pop('r', 0.5)
+    tmp = np.where(abs(x[1]) >= 1e-6, -x[1] / abs(x[1]) * (r - abs(x[0])), 0)
+    return tmp * (abs(x[0]) < r) * (abs(x[1]) < r)
+
+
+def disp_field0(x):
+    """First component of analytic displacement field (x,y) -> eps * (x+y)."""
+    eps = 0.25
+    return eps * (x[0] + x[1])
+
+
+def disp_field1(x):
+    """Second component of analytic displacement field (x,y) -> eps * 3xy."""
+    eps = 0.25
+    return eps * 3 * x[0] * x[1]
+
+
+def deform_hat(x, **kwargs):
+    """Analytic deformed hat function with support in square with side r.
+    The analytic displacement field (x,y) -> eps * (x+y, 3xy)."""
+    r = kwargs.pop('r', 0.5)
+    eps = 0.25
+    x0 = x[0] + eps * (x[0] + x[1])
+    x1 = x[1] + eps * 3 * x[0] * x[1]
+    return (r - abs(x0)) * (abs(x0) < r) * (r - abs(x1)) * (abs(x1) < r)
+
+
+def vector_field0(x):
+    """First component of analytic displacement field (x,y) -> x-y."""
+    return x[0] - x[1]
+
+
+def vector_field1(x):
+    """Second component of analytic displacement field (x,y) -> 4xy."""
+    return 4 * x[0] * x[1]
+
+
+def hat_deform_grad0(x, **kwargs):
+    """First component of the deformed gradient of the hat function,
+    which is the same as evaluating the gradient
+    at (x, y) + eps * (2x+y, y+3xy).
+    """
+    r = kwargs.pop('r', 0.5)
+    eps = 0.25
+    x0 = x[0] + eps * (x[0] + x[1])
+    x1 = x[1] + eps * 3 * x[0] * x[1]
+
+    tmp = np.where(abs(x0) >= 1e-6, -x0 / abs(x0) * (r - abs(x1)), 0)
+    return tmp * (abs(x0) < r) * (abs(x1) < r)
+
+
+def hat_deform_grad1(x, **kwargs):
+    """Second component of the deformed gradient of the hat function,
+    which is the same as evaluating the gradient
+    at (x, y) + eps * (2x+y, y+3xy).
+    """
+    r = kwargs.pop('r', 0.5)
+    eps = 0.25
+    x0 = x[0] + eps * (x[0] + x[1])
+    x1 = x[1] + eps * 3 * x[0] * x[1]
+
+    tmp = np.where(abs(x1) >= 1e-6, -x1 / abs(x1) * (r - abs(x0)), 0)
+    return tmp * (abs(x0) < r) * (abs(x1) < r)
+
+
+def fixed_templ_Deriv_hat(x, **kwargs):
+    r = kwargs.pop('r', 0.5)
+    eps = 0.25
+    x0 = x[0] + eps * (x[0] + x[1])
+    x1 = x[1] + eps * 3 * x[0] * x[1]
+
+    tmp1 = np.where(abs(x0) >= 1e-6, -x0 / abs(x0) * (r - abs(x1)), 0)
+    dg0 = tmp1 * (abs(x0) < r) * (abs(x1) < r)
+
+    tmp2 = np.where(abs(x1) >= 1e-6, -x1 / abs(x1) * (r - abs(x0)), 0)
+    dg1 = tmp2 * (abs(x0) < r) * (abs(x1) < r)
+
+    v0 = x[0] - x[1]
+    v1 = 4 * x[0] * x[1]
+    return dg0 * v0 + dg1 * v1
+
+
+def fixed_templ_adj_hat1(x, **kwargs):
```

Comment:
Also, you sometimes use 0 indexing (`vector_field0`/`vector_field1`), and sometimes 1 indexing, (`fixed_templ_adj_hat1`/`fixed_templ_adj_hat2`). This needs to be unified.